File: Java Desktop Runtime/jruntime/ParticleSystem.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package jruntime;

/**
 *
 * @author Riju
 */

import java.util.*;

  public class ParticleSystem
  {
        ArrayList particle_list;
        Scene sceneHandle;
        
        public ParticleSystem( Scene sceneHandle )
        {
            particle_list = new ArrayList();
            this.sceneHandle = sceneHandle;
        }

        public void addParticle(Particle particle)
        {
            while (!sceneHandle.loadGameObject(particle))
            {
                particle.instance_name += ( (particle.pos_y + 10) + (new Random()).nextInt(particle.pos_y * 10) + (particle.pos_x + 10) + (new Random()).nextInt(particle.pos_x * 10));
            }

            particle_list.add(particle);
        }

        public void addParticle(String instance_name, String object_name, int pos_x, int pos_y,int direction,int speed)
        {
            Particle particle = new Particle();

            particle.instance_name = "@XZX_PARTICLE_" + ( (particle.pos_y + 10) + (new Random()).nextInt(particle.pos_y * 10) + (particle.pos_x + 10) + (new Random()).nextInt(particle.pos_x * 10)); 
            particle.obj_instance = ObjectManager.findGameObjectWithName(object_name);
            particle.pos_x = pos_x;
            particle.pos_y = pos_y;
            particle.direction = direction;
            particle.speed = speed;

            while (!sceneHandle.loadGameObject(particle))
            {
                particle.instance_name += ( (particle.pos_y + 10) + (new Random()).nextInt(particle.pos_y * 10) + (particle.pos_x + 10) + (new Random()).nextInt(particle.pos_x * 10));
            }

            particle_list.add( particle );  
        }

        public void updateSystem()
        {
            boolean hasDeleted = false;
            
            do
            {
                hasDeleted = false;
                
            for (int cnt = 0; cnt < particle_list.size(); cnt++)
            {
                if ( ((Particle) particle_list.get(cnt)).pos_x < 0 || ((Particle) particle_list.get(cnt)).pos_x > HApplication.getWindowHandle().getWidth( ) || ((Particle) particle_list.get(cnt)).pos_y < 0 || ((Particle) particle_list.get(cnt)).pos_y > HApplication.getWindowHandle().getHeight( ))
                {
                    sceneHandle.destroyGameObject(((Particle) particle_list.get(cnt)).instance_name);
                    particle_list.remove(cnt);
                    hasDeleted = true;
                    break;
                }
                else
                {
                    if (((Particle) particle_list.get(cnt)).scriptsEmpty( ))
                    {
                        if (((Particle) particle_list.get(cnt)).direction == 1)
                        {
                            ((Particle) particle_list.get(cnt)).pos_y -= ((Particle) particle_list.get(cnt)).speed; // Up
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 2)
                        {
                            ((Particle) particle_list.get(cnt)).pos_x -= ((Particle) particle_list.get(cnt)).speed; // Left
                            ((Particle) particle_list.get(cnt)).pos_y -= ((Particle) particle_list.get(cnt)).speed; // Up
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 3)
                        {
                            ((Particle) particle_list.get(cnt)).pos_x -= ((Particle) particle_list.get(cnt)).speed; // Left
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 4)
                        {
                            ((Particle) particle_list.get(cnt)).pos_x -= ((Particle) particle_list.get(cnt)).speed; // Left
                            ((Particle) particle_list.get(cnt)).pos_y += ((Particle) particle_list.get(cnt)).speed; // Down 
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 5)
                        {
                            ((Particle) particle_list.get(cnt)).pos_y += ((Particle) particle_list.get(cnt)).speed; // Down
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 6)
                        {
                            ((Particle) particle_list.get(cnt)).pos_x += ((Particle) particle_list.get(cnt)).speed; // Right
                            ((Particle) particle_list.get(cnt)).pos_y += ((Particle) particle_list.get(cnt)).speed; // Down
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 7)
                        {
                            ((Particle) particle_list.get(cnt)).pos_x += ((Particle) particle_list.get(cnt)).speed; // Right
                        }
                        else if (((Particle) particle_list.get(cnt)).direction == 8)
                        {
                            ((Particle) particle_list.get(cnt)).pos_x += ((Particle) particle_list.get(cnt)).speed; // Right
                            ((Particle) particle_list.get(cnt)).pos_y -= ((Particle) particle_list.get(cnt)).speed; // Up.
                        }
                    }
                }
            }
            
            } while (hasDeleted);
        }
 }

File: C++ Linux Runtime/ParticleSystem.cpp
#include "ParticleSystem.h"
#include "ObjectManager.h"
#include "HApplication.h"

ParticleSystem::ParticleSystem(Scene_ptr scene)
{
	this->sceneHandle = scene;
}

void ParticleSystem::addParticle(Particle * particle_ptr)
{
	while (!sceneHandle->loadGameObject(particle_ptr))
	{
		particle_ptr->instance_name += ((particle_ptr->pos_x + 10) + rand() % (particle_ptr->pos_x * 10)) + ((particle_ptr->pos_y + 10) + rand() % (particle_ptr->pos_y * 10));
	}

	particle_list.push_back(particle_ptr);
}

void ParticleSystem::addParticle(std::string instance_name, std::string object_name, int pos_x, int pos_y, int direction, int speed)
{
	Particle * particle_ptr = new Particle();

	particle_ptr->instance_name = "@XZX_PARTICLE_" + ((particle_ptr->pos_x + 10) + rand() % (particle_ptr->pos_x * 10)) + ((particle_ptr->pos_y + 10) + rand() % (particle_ptr->pos_y * 10)) + instance_name;
	particle_ptr->obj_instance = ObjectManager::findGameObjectWithName(object_name);
	particle_ptr->pos_x = pos_x;
	particle_ptr->pos_y = pos_y;
	particle_ptr->direction = direction;
	particle_ptr->speed = speed;

	while (!sceneHandle->loadGameObject(particle_ptr))
	{
		particle_ptr->instance_name += ((particle_ptr->pos_x + 10) + rand() % (particle_ptr->pos_x * 10)) + ((particle_ptr->pos_y + 10) + rand() % (particle_ptr->pos_y * 10));
	}

	particle_list.push_back(particle_ptr);
}

void ParticleSystem::updateSystem()
{
re:

	for (int cnt = 0; cnt < particle_list.size( ); cnt++)
	{
		Particle_ptr particle_ptr = NULL;
		int cntr = 0;

		for (std::list<Particle_ptr>::iterator cur_particle = particle_list.begin(); cur_particle != particle_list.end(); cur_particle++,cntr++)
		{
			if (cnt == cntr)
			{
				particle_ptr = *cur_particle;
			}
		}

		SDL_Rect base_rect;

		SDL_GetWindowSize(HApplication::getWindowHandle(), &base_rect.w, &base_rect.h);

		if (particle_ptr->pos_x < 0 || particle_ptr->pos_x > base_rect.w || particle_ptr->pos_y < 0 || particle_ptr->pos_y > base_rect.h)
		{
			sceneHandle->destroyGameObject(particle_ptr->instance_name);

			cntr = 0;

			for (std::list<Particle_ptr>::iterator cur_particle = particle_list.begin(); cur_particle != particle_list.end(); cur_particle++, cntr++)
			{
				if (cnt == cntr)
				{
					particle_list.erase(cur_particle);
				}
			}

			goto re;
		}
		else
		{
			if (particle_ptr->scriptsEmpty())
			{
				if (particle_ptr->direction == 1)
				{
					particle_ptr->pos_y -= particle_ptr->speed; // Up
				}
				else if (particle_ptr->direction == 2)
				{
					particle_ptr->pos_x -= particle_ptr->speed; // Left
					particle_ptr->pos_y -= particle_ptr->speed; // Up
				}
				else if (particle_ptr->direction == 3)
				{
					particle_ptr->pos_x -= particle_ptr->speed; // Left
				}
				else if (particle_ptr->direction == 4)
				{
					particle_ptr->pos_x -= particle_ptr->speed; // Left
					particle_ptr->pos_y += particle_ptr->speed; // Down 
				}
				else if (particle_ptr->direction == 5)
				{
					particle_ptr->pos_y += particle_ptr->speed; // Down
				}
				else if (particle_ptr->direction == 6)
				{
					particle_ptr->pos_x += particle_ptr->speed; // Right
					particle_ptr->pos_y += particle_ptr->speed; // Down
				}
				else if (particle_ptr->direction == 7)
				{
					particle_ptr->pos_x += particle_ptr->speed; // Right
				}
				else if (particle_ptr->direction == 8)
				{
					particle_ptr->pos_x += particle_ptr->speed; // Right
					particle_ptr->pos_y -= particle_ptr->speed; // Up.
				}
			}
		}
	}
}